lib/openpgp.php

```diff
@@ -705,9 +705,18 @@ function __construct($data=NULL, $key_algorithm=NULL, $hash_algorithm=NULL) {
   function sign_data($signers) {
     $this->trailer = $this->calculate_trailer();
     $signer = $signers[$this->key_algorithm_name()][$this->hash_algorithm_name()];
-    $this->data = call_user_func($signer, $this->data.$this->trailer);
-    $unpacked = unpack('n', substr(implode('',$this->data), 0, 2));
+    $string_to_sign = $this->data.$this->trailer;
+
+    try {
+      $hasher = new \phpseclib3\Crypt\Hash($this->hash_algorithm_name());
+      $hash = $hasher->hash($string_to_sign);
+    } catch(\phpseclib3\Exception\UnsupportedAlgorithmException $e) {
+      $hash = $string_to_sign;
+    }
+
+    $unpacked = unpack('n', substr($hash, 0, 2));
     $this->hash_head = reset($unpacked);
+    $this->data = call_user_func($signer, $string_to_sign);
```

Comment:
We want to send `$string_to_sign` to the `$signer` callbacks. They handle the hashing with phpseclib. We generate a duplicate here because we need a signature against the hash value we can't access otherwise due to encapsulation. We only store the first 2 bytes of that hash in `hash_head`.

If `hash_head` is invalid, it's enough to have PGP keys and messages marked "INVALID" by projects using the RNP library (e.g. Thunderbird). In the case of Thunderbird, it refuses to import keys or decrypt messages.